docs/Windowing.md

```diff
@@ -266,3 +266,108 @@ tuples can be received within the timeout period.
+The state that is saved includes the tuples in the window, any system state that is required to recover the state of processing
+and also the user state.
+
+```java
+topologyBuilder.setBolt("mybolt",
+                   new MyStatefulPersistentWindowedBolt()
+                   .withWindow(...) // windowing configuarations
+                   .withPersistence() // persist the window state
+                   .withMaxEventsInMemory(25000), // max number of events to be cached in memory
+                    parallelism)
+               .shuffleGrouping("spout");
+
+```
+
+The `withPersistence` instructs the framework to transparently save the tuples in window along with
+any associated system and user state to the state backend. The `withMaxEventsInMemory` is an optional 
+configuration that specifies the maximum number of tuples that may be kept in memory. The tuples are transparently loaded from 
+the state backend as required and the ones that are most likely to be used again are retained in memory.
+
+The state backend can be configured by setting the topology state provider config,
+
+```java
+// use redis for state persistence
+conf.put(Config.TOPOLOGY_STATE_PROVIDER, "org.apache.storm.redis.state.RedisKeyValueStateProvider");
+
+```
+Currently storm supports Redis and HBase as state backends and uses the underlying state-checkpointing
+framework for saving the window state. For more details on state checkpointing see [State-checkpointing.md](State-checkpointing.md)
+
+Here is an example of a persistent windowed bolt that uses the window checkpointing to save its state. The `initState`
+is invoked with the last saved state (user state) at initialization time. The execute method is invoked based on the configured
+windowing parameters and the tuples in the active window can be accessed via an `iterator` as shown below.
+
+```java
+public class MyStatefulPersistentWindowedBolt extends BaseStatefulWindowedBolt<K, V> {
+  private KeyValueState<K, V> state;
+  
+  @Override
+  public void initState(KeyValueState<K, V> state) {
+    this.state = state;
+   // ...
+   // restore the state from the last saved state.
+   // ...
+  }
+  
+  @Override
+  public void execute(TupleWindow window) {      
+    // iterate over tuples in the current window
+    Iterator<Tuple> it = window.getIter();
+    while (it.hasNext()) {
+        // compute some result based on the tuples in window
+    }
+    
+    // possibly update any state to be maintained across windows
+    state.put(STATE_KEY, updatedValue);
+    
+    // emit the results downstream
+    collector.emit(new Values(result));
+  }
+}
+```
+
+Note: In case of persistent windowed bolts, use `TupleWindow.getIter` to retrieve an iterator over the 
+events in the window. If the number of tuples in windows is huge, invoking `TupleWindow.get` would
+try to load all the tuples into memory and may throw an OOM exception.
```

Comment:
Do we also want to say `getNew` and `getExpired` are not supported with persistent windowed bolts?